crates/codegen/runtime/cargo/src/runtime/napi_interface/parse_error.rs

```diff
@@ -20,11 +21,13 @@ impl From<RustParseError> for ParseError {
 impl ParseError {
     #[napi(getter, ts_return_type = "text_index.TextRange", catch_unwind)]
     pub fn text_range(&self) -> TextRange {
-        self.0.text_range().into()
+        self.0.text_range().clone().into()
     }
 
-    #[napi(catch_unwind)]
-    pub fn to_error_report(&self, source_id: String, source: String, with_color: bool) -> String {
-        self.0.to_error_report(&source_id, &source, with_color)
+    #[napi(ts_return_type = "diagnostic.Diagnostic", catch_unwind)]
+    pub fn to_diagnostic(&self) -> Diagnostic {
+        // TODO: Figure out if we can auto-gen Diagnostics methods
```

Comment:
I'm somewhat torn here; I would prefer to already have the Diagnostic methods already exposed for generated `class ParseError` since the underlying Rust type already implements the trait (which is exposed as `interface Diagnostic` on the JS side) but I'd rather not do that manually for all future errors that implement diagnostic, so I've settled on the conversion for the user to be aware that this type exists and is a main type expected to be used in error reporting, regardless of their formatter choice.

Reply:
I think `Diagnostic` is exposed to TS here as a `class` not an `interface`. Is this intended?

I think this is part of a bigger discussion of whether we want the general API to return the generalized type (`dyn Diagnostic` in Rust and `Diagnostic` interface types in TS), or use a more typed approach (ADTs in Rust and some form of discriminated unions in TS) to always expose the concrete types.

Reply:
> Is this intended?

That's how napi-rs works for tuple structs. We expose an opaque value that has a public interface, so they decided to model it with a `class` IIUC. Plain objects/POD are implemented by them as an `interface` with the relevant public fields. If you know how we might be consistent and use interfaces all around, then I'm open.

Unfortunately, napi-rs probably can't technically analyze the type system to see which value implements which trait, which it could in theory. model with an `interface`, because it would require type knowledge and keeping a database when executing the proc macro to know which fields of the trait should be put in the implementer's interface. That means, my `TODO` comment is probably wishful thinking but I'm surprised no-one has bumped into this and/or has implemented some hacky solution inside the `napi-rs`.

Re typed approach, this is rather a discussion how we might surface errors in general but this PR is about the interface for surfacing diagnostic API for editor integration; let's keep it about the API and we can talk about the errors separately.

Lastly, I think the terminology is clear here: we have "parse errors" that result in "(compiler) diagnostics" (surfaced to the user/editor); I don't think it's necessary to special-case and split compiler/parser and have "parser diagnostics" here. Let's keep it simple :+1: